alembic/toctree.rst

```diff
@@ -1,7 +1,9 @@
 ********************************
 Database migration using Alembic
 ********************************
-:index:`docs to write`: overall approach/use. This is a generic, single-database configuration.
+:index:`docs to write`: **overall approach/use.** This is a generic, single-database configuration.
+
+:index:`TODO`: **What's the plan for migrations?** web2py also does migrations, so how do we avoid a migration duel between web2py and Alembic?
```

Comment:
Great question.  What I love about alembic is that migrations are intentional, both in their creation and in their applicaition.  Whereas web2py wants to do migrations automatically. ( I have them disabled for production and development and use `rsmanage migrate` myself)

We have definitely added some extra work for ourselves here, but I don't know if we can avoid it as we are building for two important use cases.

1. A new book server for our large live public systems (runestone.academy)
2. An easy to use replacement for the current `runestone serve`

If all we cared about was  1, we could just have sqlalchemy slurp up the table definitions from the database.  But that does not work for 2.  

I think the way we need to proceed is to be very careful about schema changes and when they are necessary we make sure we do them in both places.

Reply:
I definitely prefer the idea of a deliberate, manually-invoked migration over web2py's "just do it" approach. Some thoughts and questions:

- I agree with the goal of replacing `runestone serve` in the long run, but I think we should first get the absolute minimum of endpoints working then deploy that before moving forward. IMHO, what you're currently started (`hsblog`, `getAssessResults`, and book serving) should represent a vast majority of the site traffic. I propose not adding any more endpoints until this is in production.
- This means clearly diving the database migration into "managed by Alembic" (all tables SQLAlchemy knows about) and "managed by web2py" (everything else). Then, after struggling through a migration, we need to write up a guide on how to do this correctly. My guess:
    - Update model definitions in both web2py and BookServer.
    - Shut down web2py.
    - Migrate via Alembic.
    - Re-enable web2py